Additions to AWS Module Guidlines from Pull Request Feedback

orthanc opened the pull request:
##### SUMMARY
A number of documentation updates to the AWS module guidelines and AnsibleAWSModule suport class to capture things that I've only been able to find from copying other modules and pull request feedback or otherwise have tripped me up a couple of times.

Specifically:
* New AnsibleAWSModule connection helpers`client` and `resource`
* Requirement for integration tests for new modules
* Necessary aliases file to get AWS credentials in integration tests
* How to access aws credentials in integration test
* Need to update permissions for integration tests
* Common documentation fragments that should be used

I've also removed the try/except block from the documentation in AnsibleAWSModule connect since according to the guidelines these methods do not throw connection exceptions and the exceptions instead come when using the connection.

Happy to move or restructure if there is a more appropriate place for any of this information.

##### ISSUE TYPE
 - Docs Pull Request

##### COMPONENT NAME
aws

##### ANSIBLE VERSION
```
ansible 2.6.0 (docfix_connect_in_ansible_aws_module 41bee9fda3) last updated 2018/02/28 22:19:52 (GMT +1300)
  config file = None
  configured module search path = [u'USER_HOME/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = ANSIBLE_CHECKOUT/lib/ansible
  executable location = ANSIBLE_CHECKOUT/bin/ansible
  python version = 2.7.14 (default, Sep 23 2017, 22:06:14) [GCC 7.2.0]
```


##### ADDITIONAL INFORMATION

Most of these are triggered by feedback on #36683

orthanc committed:
Update doc of AnsibleAWSModule to remove incorrect connect example.

Current example uses aws_connect method which no longer exists. Replace
this with the client and resource methods that do exist.

Also remove try/except block in connect example as guidelines imply that
there aren't connection errors thrown on connection, just when later
sing the connection.

orthanc committed:
Update AWS module guidelines to include the AnsibleAWSModule connection methods.

orthanc committed:
Add information on integration testing to the AWS module guidelines.

orthanc committed:
Add information on common documentation fragments to AWS module
Guidelines.

willthames commented on lib/ansible/modules/cloud/amazon/GUIDELINES.md:
````diff
@@ -496,3 +531,63 @@ other helper function `boto3_tag_list_to_ansible_dict` to get an appropriate tag
+
+```yaml
+  - ec2_instance:
+      ... params ...
````
should be updated to use YAML anchors

willthames commented on lib/ansible/modules/cloud/amazon/GUIDELINES.md:
````diff
@@ -158,11 +159,45 @@ caught, you should catch `ClientError` and `BotoCoreError` exceptions with every
 See exception handling. module_utils.ec2 checks for missing profiles or a region not set when it needs to be,
 so you don't have to.
 
+```python
````
This is very cool, I'd missed this somehow. Can we move this nearer to your new line 129 to make it more obvious?

willthames commented on lib/ansible/modules/cloud/amazon/GUIDELINES.md:
````diff
@@ -496,3 +531,63 @@ other helper function `boto3_tag_list_to_ansible_dict` to get an appropriate tag
+
+```
+cloud/aws
+posix/ci/cloud/group4/aws
````
group4 isn't a must - it should be one in of the (currently) 5 groups though

orthanc replied:
Is there a rationale I should include for how to pick a group? Or is it just as simple as "we want to distribute the tests between the groups, doesn't really matter which group any given test is in" ?

willthames replied:
I struggle to answer that. I think I should know how to work that out (put it in the group that's currently taking the last time) but I don't know how to get that information. 

Perhaps @mattclay or @s-hertel could help answer. Your wording would certainly work in the absence of a more definitive answer.

mattclay replied:
I recommend stating that new AWS tests should use the same group as existing AWS tests. If we can avoid a direct reference to group4 that will mean one less place to update documentation when it changes.

willthames commented:
This is great work, thanks @orthanc.

orthanc committed:
Changes as requested on PR.

* Restructure connection section to start with current best practice
* Explain the use of the CI build groups
* Use YAML Anchors for aws credentials example

mattclay commented on lib/ansible/modules/cloud/amazon/GUIDELINES.md:
````diff
@@ -496,3 +531,85 @@ other helper function `boto3_tag_list_to_ansible_dict` to get an appropriate tag
+
+```
+cloud/aws
+posix/ci/cloud/group2/aws
````
Contributors should be instructed to use the same group as existing AWS tests, which happens to be group4 currently. If we avoid referencing the group directly then it won't be out of date when the grouping changes.

orthanc replied:
Thinking about how to make this clear. I think what I'll do is word it to be existing tests and link to one of the existing tests rather than including an inline example.

That way it's obvious what to copy without creating the documentation update overhead. Does that work?

mattclay replied:
That works for me.

orthanc committed:
Replace guidance on use of test groups with link to aliases file.

This achieves the goal of explaining why this file is necessary without
introducing overhead of keeping documentation up to date when test
groups change.

willthames commented:
@orthanc I do have a reason not to reorder the boto/boto3 stuff -  I have a separate PR to eliminate the boto information entirely.

I'm happy to rebase #36763 on top of yours but would rather that be as little effort as possible.

orthanc commented:
@willthames that's fine, I've removed the reordering commit. If the intention is to remove this info anyway there's no point reordering.

orthanc commented:
Just rebasing on top of current devel to re-trigger the build since there was an unrelated failure. No actual changes since willthames review.

mattclay commented:
@orthanc For future reference, when there is an unrelated failure, simply closing and re-opening the PR will trigger CI. A rebase is not needed unless there are merge conflicts with your PR and devel.

willthames commented:
Merged, thanks again @orthanc
